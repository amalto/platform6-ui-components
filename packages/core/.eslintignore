dist/
loader/

src/components.d.ts