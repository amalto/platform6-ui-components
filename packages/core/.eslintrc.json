{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
