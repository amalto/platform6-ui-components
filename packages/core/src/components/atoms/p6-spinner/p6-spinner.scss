:host {
  display: inline-flex;
  justify-content: center;
  align-items: center;

  rect[class^='fix-moz-svg-anime-'] {
    animation-name: fadein;
    animation-iteration-count: infinite;
    animation-duration: 0.84s;
  }

  @for $i from 1 through 7 {
    .fix-moz-svg-anime-#{$i} {
      animation-delay: 0.105s * $i;
    }
  }

  @keyframes fadein {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }
}
