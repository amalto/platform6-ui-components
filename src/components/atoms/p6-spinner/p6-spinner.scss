:host {
    display: inline-flex;
    justify-content: center;
    align-items: center;

    img {
        object-fit: contain;
        object-position: center;
    }
  }