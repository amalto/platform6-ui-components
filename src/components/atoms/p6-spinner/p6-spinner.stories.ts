import { Components } from "../../../components";
import {
  getElement,
  makeStory,
  Props,
} from "../../../shared/storybook/stories";

const component = "p6-spinner";

export default {
  title: "Atoms/Spinner",
  component,
};

const getStoryField = (props?: Props<Components.P6Spinner>): HTMLElement => {
  return getElement(component, [], props);
};

export const Default = makeStory({
  builder: (): HTMLElement =>
    getStoryField({
      style: { width: "4.2rem", height: "4.2rem" } as CSSStyleDeclaration,
    }),
});

export const SmallContainer = makeStory({
  builder: (): HTMLElement =>
    getStoryField({
      style: { width: "1rem", height: "1rem" } as CSSStyleDeclaration,
    }),
});

export const VerticalContainer = makeStory({
  builder: (): HTMLElement =>
    getStoryField({
      style: {
        width: "10.5rem",
        height: "21rem",
        background: "#eee",
      } as CSSStyleDeclaration,
    }),
});

export const HorizontalContainer = makeStory({
  builder: (): HTMLElement =>
    getStoryField({
      style: {
        width: "21rem",
        height: "10.5rem",
        background: "#eee",
      } as CSSStyleDeclaration,
    }),
});
