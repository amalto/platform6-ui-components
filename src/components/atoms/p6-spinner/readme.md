# p6-spinner

<!-- Auto Generated Below -->

## Dependencies

### Used by

- [p6-grid](../../organisms/p6-grid)
- [p6-waiting](../../templates/p6-waiting)

### Graph

```mermaid
graph TD;
  p6-grid --> p6-spinner
  p6-waiting --> p6-spinner
  style p6-spinner fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
