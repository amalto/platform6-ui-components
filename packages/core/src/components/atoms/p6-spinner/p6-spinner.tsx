import { Component, h, Host, JSX } from '@stencil/core';

interface Coord {
  x: number;
  y: number;
}

@Component({
  tag: 'p6-spinner',
  styleUrl: 'p6-spinner.scss',
  shadow: true,
})
export class P6Spinner {
  private readonly color = '#fb9248';

  private readonly speed = 0.105;

  private readonly duration = 0.84;

  private readonly blockSize = 32;

  private readonly gap = (100 - 3 * this.blockSize) / 2;

  private readonly coords: Coord[] = [
    { x: 0, y: 0 },
    { x: this.blockSize + this.gap, y: 0 },
    { x: 2 * this.blockSize + this.gap * 2, y: 0 },
    { x: 2 * this.blockSize + this.gap * 2, y: this.blockSize + this.gap },
    {
      x: 2 * this.blockSize + this.gap * 2,
      y: 2 * this.blockSize + this.gap * 2,
    },
    { x: this.blockSize + this.gap, y: 2 * this.blockSize + this.gap * 2 },
    { x: 0, y: 2 * this.blockSize + this.gap * 2 },
    { x: 0, y: this.blockSize + this.gap },
  ];

  render(): JSX.Element {
    return (
      <Host>
        <svg width="100" height="100" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" aria-label="loading" aria-hidden="false" aria-busy="true">
          {this.coords.map(this.blockElement.bind(this))}
        </svg>
      </Host>
    );
  }

  private blockElement(coord: Coord, index: number): JSX.Element {
    const classes = {
      [`fix-moz-svg-anime-${index}`]: navigator.userAgent.includes('Firefox'),
    };

    return (
      <rect class={classes} x={coord.x} y={coord.y} width={this.blockSize} height={this.blockSize} fill={this.color}>
        <animate attributeName="opacity" from="1" to="0" dur={this.duration} repeatCount="indefinite" begin={this.speed * index} />
      </rect>
    );
  }
}
