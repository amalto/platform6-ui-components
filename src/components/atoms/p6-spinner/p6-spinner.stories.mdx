import { Story, Preview, Meta } from "@storybook/addon-docs/blocks";

import readme from "./readme.md";

<Meta
  title="Atoms/Spinner"
  component="p6-spinner"
  parameters={{ notes: readme }}
/>

# Spinner

<Preview>
  <Story name="Default" height="42px">
    {`
      <p6-spinner style="width: 4.2rem; height: 4.2rem"></p6-spinner>
    `}
  </Story>
</Preview>

<Preview>
  <Story name="Small container" height="42px">
    {`
      <p6-spinner style="width: 1rem; height: 1rem"></p6-spinner>
    `}
  </Story>
</Preview>

<Preview>
  <Story name="Vertical container" height="42px">
    {`
      <p6-spinner style="width: 10.5rem; height: 21rem; background: #eee"></p6-spinner>
    `}
  </Story>
</Preview>

<Preview>
  <Story name="Horizontal container" height="42px">
    {`
      <p6-spinner style="height: 10.5rem; width: 21rem; background: #eee"></p6-spinner>
    `}
  </Story>
</Preview>
