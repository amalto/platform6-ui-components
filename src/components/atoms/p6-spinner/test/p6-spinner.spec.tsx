import { newSpecPage } from '@stencil/core/testing';
import { P6Spinner } from '../p6-spinner';

describe('p6-spinner', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6Spinner],
      html: `<p6-spinner></p6-spinner>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
