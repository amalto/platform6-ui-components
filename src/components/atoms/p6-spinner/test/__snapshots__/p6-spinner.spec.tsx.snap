// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-spinner renders 1`] = `
<p6-spinner>
  <mock:shadow-root>
    <svg aria-busy="true" aria-hidden="false" aria-label="loading" height="100" preserveAspectRatio="xMidYMid" viewBox="0 0 100 100" width="100">
      <rect fill="#fb9248" height="32" width="32" x="0" y="0">
        <animate attributeName="opacity" begin="0" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="34" y="0">
        <animate attributeName="opacity" begin="0.105" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="68" y="0">
        <animate attributeName="opacity" begin="0.21" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="68" y="34">
        <animate attributeName="opacity" begin="0.315" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="68" y="68">
        <animate attributeName="opacity" begin="0.42" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="34" y="68">
        <animate attributeName="opacity" begin="0.525" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="0" y="68">
        <animate attributeName="opacity" begin="0.63" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
      <rect fill="#fb9248" height="32" width="32" x="0" y="34">
        <animate attributeName="opacity" begin="0.735" dur="0.84" from="1" repeatCount="indefinite" to="0"></animate>
      </rect>
    </svg>
  </mock:shadow-root>
</p6-spinner>
`;
